File: internal/orchestrator/panes_test.go
package orchestrator

import (
	"context"
	"errors"
	"strings"
	"testing"
	"time"
)

func rects(sizes ...[3]int) []paneRect {
	var panes []paneRect
	for i, s := range sizes {
		var p paneRect
		p.PaneID, p.Rect.Width, p.Rect.Height = string(rune('a'+i)), s[1], s[2]
		panes = append(panes, p)
	}
	return panes
}

func TestANewStagePaneComesOutOfTheRoomiestPaneAlongItsLongerSide(t *testing.T) {
	// {_, width, height} in cells; a cell is twice as tall as it is wide, so
	// 200x100 is a square on screen and 100x100 is a tall sliver.
	for _, c := range []struct {
		name            string
		panes           []paneRect
		pane, direction string
	}{
		{"a wide pane is cut down the middle", rects([3]int{0, 400, 100}), "a", "right"},
		{"a tall pane is cut across", rects([3]int{0, 100, 100}), "a", "down"},
		{"a square pane is cut across", rects([3]int{0, 200, 100}), "a", "down"},
		{"the roomiest pane is the one cut", rects([3]int{0, 100, 50}, [3]int{0, 400, 100}, [3]int{0, 100, 100}), "b", "right"},
		{"no layout, no answer", nil, "", ""},
	} {
		t.Run(c.name, func(t *testing.T) {
			pane, direction := splitTarget(c.panes)
			if pane != c.pane || direction != c.direction {
				t.Errorf("splitTarget = %q %q, want %q %q", pane, direction, c.pane, c.direction)
			}
		})
	}
}

func TestAStagePaneIsSplitTheWayTheTabIsShaped(t *testing.T) {
	for _, c := range []struct {
		name, direction string
		rect            [2]int
	}{
		{"a tall window is cut across", "down", [2]int{100, 400}},
		{"a wide window is cut down the middle", "right", [2]int{400, 50}},
	} {
		t.Run(c.name, func(t *testing.T) {
			w, o := newWorld(t, &bdTicket{ID: "hx-1"})
			w.rect = c.rect
			o.runTicket(context.Background(), "hx-1")

			split := w.Called("herdr pane split")
			if len(split) == 0 {
				t.Fatal("no pane was split")
			}
			if !strings.Contains(split[0], "--direction "+c.direction) {
				t.Errorf("a %v pane was cut the other way: %q", c.rect, split[0])
			}
		})
	}
}

func TestAPaneThatHasNotGotItsShellYetIsWaitedFor(t *testing.T) {
	w, o := newWorld(t, &bdTicket{ID: "hx-1"})
	// herdr refuses until the pane it just created has a shell.
	w.failOnce("herdr agent start", errors.New(`{"error":{"code":"agent_pane_busy","message":"agent target w1:p2 is not an available shell"}}`))

	o.runTicket(context.Background(), "hx-1")

	if got := o.ticket("hx-1"); got.Status != statusPROpen {
		t.Errorf("a pane that was a moment from ready ended the Stage: %+v", got)
	}
}

func TestASessionStillPickingUpItsPromptIsNotAFinishedStage(t *testing.T) {
	w, o := newWorld(t, &bdTicket{ID: "hx-1"})
	o.Tick = 10 * time.Millisecond // a 30ms grace for the session to get going
	w.session = func(p prompt) (string, string) {
		// Reads idle at once, as a real session does, and writes its result
		// a moment later.
		go func() {
			time.Sleep(5 * time.Millisecond)
			writeFile(t, p.file, "STATUS: done\nPR: https://example.test/pr\n")
		}()
		return "", "idle"
	}

	o.runTicket(context.Background(), "hx-1")

	for _, line := range w.mainLines() {
		if strings.Contains(line, "WAKE") {
			t.Errorf("a session that was still starting was woken on: %q", line)
		}
	}
	if got := o.ticket("hx-1"); got.Status != statusPROpen {
		t.Errorf("the Pipeline did not finish: %+v", got)
	}
}

File: src/orchestrator/panes_test.rs
use super::herdr::{split_target, PaneRect, Rect};
use super::state::STATUS_PR_OPEN;
use super::world::{new_world, BdTicket};
use super::write_file;
use std::path::PathBuf;
use std::thread;
use std::time::Duration;

fn rects(sizes: &[(usize, usize)]) -> Vec<PaneRect> {
    sizes
        .iter()
        .enumerate()
        .map(|(i, &(width, height))| PaneRect {
            pane_id: ((b'a' + i as u8) as char).to_string(),
            rect: Rect { width, height },
        })
        .collect()
}

#[test]
fn a_new_stage_pane_comes_out_of_the_roomiest_pane_along_its_longer_side() {
    // (width, height) in cells; a cell is twice as tall as it is wide, so
    // 200x100 is a square on screen and 100x100 is a tall sliver.
    let cases: [(&str, Vec<PaneRect>, &str, &str); 5] = [
        (
            "a wide pane is cut down the middle",
            rects(&[(400, 100)]),
            "a",
            "right",
        ),
        (
            "a tall pane is cut across",
            rects(&[(100, 100)]),
            "a",
            "down",
        ),
        (
            "a square pane is cut across",
            rects(&[(200, 100)]),
            "a",
            "down",
        ),
        (
            "the roomiest pane is the one cut",
            rects(&[(100, 50), (400, 100), (100, 100)]),
            "b",
            "right",
        ),
        ("no layout, no answer", Vec::new(), "", ""),
    ];
    for (name, panes, pane, direction) in cases {
        assert_eq!(
            split_target(&panes),
            (pane.to_string(), direction.to_string()),
            "{name}"
        );
    }
}

#[test]
fn a_stage_pane_is_split_the_way_the_tab_is_shaped() {
    for (name, direction, rect) in [
        ("a tall window is cut across", "down", (100, 400)),
        ("a wide window is cut down the middle", "right", (400, 50)),
    ] {
        let (w, o) = new_world(vec![BdTicket::new("hx-1")]);
        w.lock().rect = rect;
        o.run_ticket("hx-1");

        let split = w.called("herdr pane split");
        assert!(!split.is_empty(), "{name}: no pane was split");
        assert!(
            split[0].contains(&format!("--direction {direction}")),
            "{name}: a {rect:?} pane was cut the other way: {:?}",
            split[0]
        );
    }
}

#[test]
fn a_pane_that_has_not_got_its_shell_yet_is_waited_for() {
    let (w, o) = new_world(vec![BdTicket::new("hx-1")]);
    // herdr refuses until the pane it just created has a shell.
    w.fail_once(
        "herdr agent start",
        r#"{"error":{"code":"agent_pane_busy","message":"agent target w1:p2 is not an available shell"}}"#,
    );

    o.run_ticket("hx-1");

    let got = o.ticket("hx-1");
    assert_eq!(
        got.status, STATUS_PR_OPEN,
        "a pane that was a moment from ready ended the Stage: {got:?}"
    );
}

#[test]
fn a_session_still_picking_up_its_prompt_is_not_a_finished_stage() {
    let (w, mut o) = new_world(vec![BdTicket::new("hx-1")]);
    o.cfg.tick = Duration::from_millis(10); // a 30ms grace for the session to get going
    w.session(|p| {
        // Reads idle at once, as a real session does, and writes its result
        // a moment later.
        let file = PathBuf::from(&p.file);
        thread::spawn(move || {
            thread::sleep(Duration::from_millis(5));
            write_file(&file, "STATUS: done\nPR: https://example.test/pr\n");
        });
        (String::new(), "idle".to_string())
    });

    o.run_ticket("hx-1");

    for line in w.main_lines() {
        assert!(
            !line.contains("WAKE"),
            "a session that was still starting was woken on: {line:?}"
        );
    }
    let got = o.ticket("hx-1");
    assert_eq!(
        got.status, STATUS_PR_OPEN,
        "the Pipeline did not finish: {got:?}"
    );
}
